File: 031/eu031a.py
# project euler (projecteuler.net) problem 31
# solution by Kevin Retzke (retzkek@gmail.com), May 2012

def countCombinations(vals, total):
    if vals == []:
        return 1 if total == 0 else 0
    count = 0
    if vals[0] <= total:
        count += countCombinations(vals, total-vals[0])
    count += countCombinations(vals[1:], total)
    return count

if __name__=="__main__":
    coins = [200, 100, 50, 20, 10, 5, 2, 1]
    print countCombinations(coins, 200)

File: 031/eu031a.c
// project euler (projecteuler.net) problem 31
// solution by Kevin Retzke (retzkek@gmail.com), May 2012
#include <stdio.h>

int countCombinations(int *vals, int ival, int nvals, int total) 
{
	if (ival+1 == nvals) {
        return ((total%vals[ival]) == 0);
	}
	int count = 0;
	if (vals[ival] <= total) {
		count += countCombinations(vals, ival, nvals, total-vals[ival]);
	}
	count += countCombinations(vals, ival+1, nvals, total);
	return count;
}

int main()
{
	int coins[] = {200, 100, 50, 20, 10, 5, 2};
	printf("%i\n",countCombinations(coins, 0, 7, 200));
}
